// sigaa-api/dist/activity/sigaa-quiz-activity.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaActivityQuiz = void 0;

require("source-map-support/register");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class SigaaActivityQuiz {
  constructor(activityData, studentBond) {
    this.activityData = activityData;
    this.studentBond = studentBond;

    _defineProperty(this, "type", 'quiz');
  }
  /**
   * @inheritdoc
   */


  get quizId() {
    if (!this.activityData.form.postValues) throw new Error('SIGAA: Activity Quiz form without id.');
    return this.activityData.form.postValues['id'];
  }
  /**
   * @inheritdoc
   */


  get quizTitle() {
    return this.activityData.quizTitle;
  }
  /**
   * @inheritdoc
   */


  get courseId() {
    if (!this.activityData.form.postValues['idTurma']) throw new Error('SIGAA: Activity Quiz form without idTurma.');
    return this.activityData.form.postValues['idTurma'];
  }
  /**
   * @inheritdoc
   */


  get courseTitle() {
    return this.activityData.courseTitle;
  }
  /**
   * @inheritdoc
   */


  async getCourse() {
    const courses = await this.studentBond.getCourses();
    const course = courses.find(course => course.id === this.courseId);
    if (!course) throw new Error('SIGAA: Cannot find the course referenced by the activity.');
    return course;
  }
  /**
   * @inheritdoc
   */


  async getQuiz() {
    const course = await this.getCourse();
    const quizzes = await course.getQuizzes();
    const quiz = quizzes.find(quiz => quiz.id === this.quizId);
    if (!quiz) throw new Error('SIGAA: Cannot find the quiz referenced by the activity.');
    return quiz;
  }
  /**
   * @inheritdoc
   */


  get date() {
    return this.activityData.date;
  }
  /**
   * @inheritdoc
   */


  get done() {
    return this.activityData.done;
  }

}

exports.SigaaActivityQuiz = SigaaActivityQuiz;
