// sigaa-api/dist/activity/sigaa-homework-activity.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaActivityHomework = void 0;

require("source-map-support/register");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Implements ActivityHomework
 * @category Public
 */
class SigaaActivityHomework {
  constructor(activityData, studentBond) {
    this.activityData = activityData;
    this.studentBond = studentBond;

    _defineProperty(this, "type", 'homework');
  }
  /**
   * @inheritdoc
   */


  get homeworkId() {
    return this.activityData.form.postValues['id'];
  }
  /**
   * @inheritdoc
   */


  get homeworkTitle() {
    return this.activityData.homeworkTitle;
  }
  /**
   * @inheritdoc
   */


  get courseId() {
    return this.activityData.form.postValues['idTurma'];
  }
  /**
   * @inheritdoc
   */


  get courseTitle() {
    return this.activityData.courseTitle;
  }
  /**
   * @inheritdoc
   */


  async getCourse() {
    const courses = await this.studentBond.getCourses();
    const course = courses.find(course => course.id === this.courseId);
    if (!course) throw new Error('SIGAA: Cannot find the course referenced by the activity.');
    return course;
  }
  /**
   * @inheritdoc
   */


  async getHomework() {
    const course = await this.getCourse();
    const homeworkList = await course.getHomeworks();
    const homework = homeworkList.find(homework => homework.id === this.homeworkId);
    if (!homework) throw new Error('SIGAA: Cannot find the homework referenced by the activity.');
    return homework;
  }
  /**
   * @inheritdoc
   */


  get date() {
    return this.activityData.date;
  }
  /**
   * @inheritdoc
   */


  get done() {
    return this.activityData.done;
  }

}

exports.SigaaActivityHomework = SigaaActivityHomework;
