// sigaa-api/dist/activity/sigaa-exam-activity.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaActivityExam = void 0;

require("source-map-support/register");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class SigaaActivityExam {
  constructor(activityData, studentBond) {
    this.activityData = activityData;
    this.studentBond = studentBond;

    _defineProperty(this, "type", 'exam');
  }
  /**
   * @inheritdoc
   */


  get examDescription() {
    return this.activityData.examDescription;
  }
  /**
   * @inheritdoc
   */


  get courseTitle() {
    return this.activityData.courseTitle;
  }
  /**
   * @inheritdoc
   */


  async getCourse() {
    const courses = await this.studentBond.getCourses();
    const validCourses = courses.filter(course => course.title === this.courseTitle);
    if (validCourses.length === 0) throw new Error('SIGAA: No course with title specified in the found activity.');
    if (validCourses.length > 1) throw new Error('SIGAA: It was not possible to identify the course as there is more than one course with the same title specified in the activity.');
    return validCourses[0];
  }
  /**
   * @inheritdoc
   */


  async getExam() {
    const course = await this.getCourse();
    const exams = await course.getExamCalendar();
    const validExams = exams.filter(exam => exam.description === this.examDescription);
    if (validExams.length === 0) throw new Error('SIGAA: Exam description specified in activity not found.');
    if (validExams.length > 1) throw new Error('SIGAA: The exam description specified in the activity corresponds to more than one exam.');
    return validExams[0];
  }
  /**
   * @inheritdoc
   */


  get date() {
    return this.activityData.date;
  }
  /**
   * @inheritdoc
   */


  get done() {
    return this.activityData.done;
  }

}

exports.SigaaActivityExam = SigaaActivityExam;
