// sigaa-api/dist/courses/sigaa-course-resources-manager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CourseResourcesManager = void 0;

require("source-map-support/register");

var _sigaaResourceManager = require("../resources/sigaa-resource-manager");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Class that stores course resources.
 * @category Internal
 */
class CourseResourcesManager {
  constructor(http, courseResourcesFactory, course) {
    _defineProperty(this, "lessons", void 0);

    _defineProperty(this, "files", void 0);

    _defineProperty(this, "homework", void 0);

    _defineProperty(this, "forums", void 0);

    _defineProperty(this, "quizzes", void 0);

    _defineProperty(this, "survey", void 0);

    _defineProperty(this, "webContents", void 0);

    _defineProperty(this, "news", void 0);

    this.lessons = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createLessonFromLessonData(options, http, async () => {
      await course.getLessons();
    }));
    this.files = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createFileFromFileData(options, http, async () => {
      await course.getFiles();
    }));
    this.homework = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createHomeworkFromHomeworkData(options, http, async () => {
      await course.getHomeworks();
    }));
    this.forums = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createForumFromForumData(options, http, async () => {
      await course.getForums();
    }));
    this.quizzes = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createQuizFromQuizData(options, http, async () => {
      await course.getQuizzes();
    }));
    this.news = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createNewsFromNewsData(options, http, async () => {
      await course.getNews();
    }));
    this.webContents = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createWebContentFromWebContentData(options, http, async () => {
      await course.getWebContents();
    }));
    this.survey = new _sigaaResourceManager.ResourceManager(options => courseResourcesFactory.createSurveyFromSurveyData(options, http, async () => {
      await course.getSurveys();
    }));
  }

}

exports.CourseResourcesManager = CourseResourcesManager;
