// sigaa-api/dist/courses/resources/sigaa-lesson-student.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaLesson = void 0;

require("source-map-support/register");

var _updatableResource = require("../../resources/updatable-resource");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @category Internal
 */
class SigaaLesson extends _updatableResource.AbstractUpdatableResource {
  constructor(options, updater) {
    super(options.instanceIndentifier, updater);

    _defineProperty(this, "_title", void 0);

    _defineProperty(this, "_id", void 0);

    _defineProperty(this, "_contextText", void 0);

    _defineProperty(this, "_startDate", void 0);

    _defineProperty(this, "_endDate", void 0);

    _defineProperty(this, "_attachments", void 0);

    this.update(options);
  }

  update(options) {
    this._title = options.title;
    this._id = options.id;
    this._contextText = options.contentText;
    this._startDate = options.startDate;
    this._endDate = options.endDate;
    this._attachments = options.attachments;
    this.isClosed = false;
  }

  get title() {
    this.checkIfItWasClosed();
    return this._title;
  }

  get id() {
    this.checkIfItWasClosed();
    return this._id;
  }

  get contentText() {
    this.checkIfItWasClosed();
    return this._contextText;
  }

  get endDate() {
    this.checkIfItWasClosed();
    return this._endDate;
  }

  get startDate() {
    this.checkIfItWasClosed();
    return this._startDate;
  }

  get attachments() {
    this.checkIfItWasClosed();
    return this._attachments;
  }

}

exports.SigaaLesson = SigaaLesson;
