// sigaa-api/dist/courses/resources/sigaa-news-student.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaNews = void 0;

require("source-map-support/register");

var _updatableResource = require("../../resources/updatable-resource");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @category Internal
 */
class SigaaNews extends _updatableResource.AbstractUpdatableResource {
  constructor(http, parser, options, updater) {
    super(options.instanceIndentifier, updater);
    this.http = http;
    this.parser = parser;

    _defineProperty(this, "_form", void 0);

    _defineProperty(this, "_title", void 0);

    _defineProperty(this, "_content", void 0);

    _defineProperty(this, "_date", void 0);

    this.update(options);
  }

  update(newsOptions) {
    this._title = newsOptions.title;
    this._form = newsOptions.form;
    this.isClosed = false;
  }

  get id() {
    this.checkIfItWasClosed();
    return this._form.postValues.id;
  }

  get title() {
    this.checkIfItWasClosed();
    return this._title;
  }

  async getContent() {
    this.checkIfItWasClosed();

    if (this._content === undefined) {
      await this.getFullNews();
    }

    return this._content;
  }

  async getDate() {
    this.checkIfItWasClosed();

    if (this._date === undefined) {
      await this.getFullNews();
    }

    return this._date;
  }

  async getFullNews(retry = true) {
    try {
      const page = await this.http.post(this._form.action.href, this._form.postValues);

      if (page.statusCode !== 200) {
        throw new Error('SIGAA: Invalid status code at news page.');
      }

      const newsElement = page.$('ul.form');

      if (newsElement.length === 0) {
        throw new Error('SIGAA: Invalid news page.');
      }

      const els = newsElement.find('span');
      const dateString = this.parser.removeTagsHtml(els.eq(1).html());
      this._date = this.parser.parseDates(dateString, 1)[0];
      this._content = this.parser.removeTagsHtml(newsElement.find('div').html());
    } catch (err) {
      if (retry) {
        await this.updateInstance();
        return this.getFullNews(false);
      } else {
        throw err;
      }
    }
  }

}

exports.SigaaNews = SigaaNews;
