// sigaa-api/dist/account/sigaa-account-factory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaAccountFactory = void 0;

require("source-map-support/register");

var _sigaaAccountIfsc = require("./sigaa-account-ifsc");

var _sigaaAccountUfpb = require("./sigaa-account-ufpb");

/**
 *
 * Serves to create account instances.
 * @category Internal
 */
class SigaaAccountFactory {
  constructor(http, parser, session, bondFactory) {
    this.http = http;
    this.parser = parser;
    this.session = session;
    this.bondFactory = bondFactory;
  }
  /**
   * Creates a new instance of Account.
   * @param page home page of account (page after login).
   */


  async getAccount(page) {
    if (this.session.institution === 'UFPB') {
      return new _sigaaAccountUfpb.SigaaAccountUFPB(page, this.http, this.parser, this.session, this.bondFactory);
    } else {
      return new _sigaaAccountIfsc.SigaaAccountIFSC(page, this.http, this.parser, this.session, this.bondFactory);
    }
  }

}

exports.SigaaAccountFactory = SigaaAccountFactory;
