// sigaa-api/dist/resources/sigaa-resource-manager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ResourceManager = void 0;

require("source-map-support/register");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @category Internal
 */
class ResourceManager {
  constructor(instanceConstructor) {
    this.instanceConstructor = instanceConstructor;

    _defineProperty(this, "_instances", []);
  }
  /**
   * Current instances.
   */


  /**
   * Closes and removes the instance if not in idsToKeep.
   * @param idsToKeep array with ids to keep E.g. ["1234", "4321"]
   */
  keepOnly(idsToKeep) {
    this._instances = this._instances.filter(instance => {
      try {
        if (idsToKeep.includes(instance._instanceIndentifier)) {
          return true;
        } else {
          instance.close();
          return false;
        }
      } catch (err) {
        return false;
      }
    });
    return this._instances;
  }
  /**
   * @inheritdoc
   */


  get instances() {
    return this._instances;
  }
  /**
   * Update instance with new information
   * If there is an instance with the instanceIndentifier equal to
   * options.instanceIndentifier, the update method will be called with
   * instanceOptions.
   * E.g. instance.update(options.instanceOptions)
   * or create new instance with constructor.
   * @param options Object with new informations
   * @return return the instance updated/created
   */


  upsert(options) {
    const id = options.instanceIndentifier;

    const instance = this._instances.find(classItem => id === classItem._instanceIndentifier);

    if (!instance) {
      const newInstance = this.instanceConstructor(options);

      this._instances.push(newInstance);

      return newInstance;
    } else {
      instance.update(options);
      return instance;
    }
  }

  close() {
    for (const instance of this.instances) {
      instance.close();
    }

    this._instances = [];
  }

}

exports.ResourceManager = ResourceManager;
