// sigaa-api/dist/resources/updatable-resource.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AbstractUpdatableResource = void 0;

require("source-map-support/register");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @category Internal
 */
class AbstractUpdatableResource {
  constructor(__instanceIndentifier, updater) {
    this.__instanceIndentifier = __instanceIndentifier;
    this.updater = updater;

    _defineProperty(this, "isClosed", false);
  }

  async updateInstance() {
    if (!this.updater) throw new Error('SIGAA: Resource updater not exists.');
    await this.updater();
  }

  checkIfItWasClosed() {
    if (this.isClosed) {
      throw new Error('SIGAA: This instance has already been closed.');
    }
  }

  get _instanceIndentifier() {
    return this.__instanceIndentifier;
  }

  close() {
    this.isClosed = false;
  }

}

exports.AbstractUpdatableResource = AbstractUpdatableResource;
