// sigaa-api/dist/helpers/sigaa-promise-stack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaPromiseStack = void 0;

require("source-map-support/register");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Class to control promise order
 * Performs chain functions and waits for each promise that the functions return.
 * @category Internal
 */
class SigaaPromiseStack {
  /**
   * Order type.
   * If it is reverse, the last entered will be the first executed. default is normal.
   */

  /**
   * Current promise running object as {key, promiseFunction, promise}.
   */

  /**
   * store all promises objects as {key, promiseFunction, promise}
   */

  /**
   * @param PromiseStackOrder [order] order of execution of the promises, if it is reverse, the last entered will be the first executed. default is normal.
   */
  constructor(order) {
    _defineProperty(this, "order", void 0);

    _defineProperty(this, "promiseRunning", void 0);

    _defineProperty(this, "storedPromises", []);

    this.order = order || 'normal';
  }
  /**
   * @inheritdoc
   */


  flush() {
    this.storedPromises = [];
    this.promiseRunning = undefined;
  }
  /**
   * @inheritdoc
   */


  get promises() {
    if (this.promiseRunning) {
      return [this.promiseRunning, ...this.storedPromises];
    } else {
      return this.storedPromises;
    }
  }
  /**
   * Loop to execute the entire promise stack
   */


  async promiseExecutor() {
    if (!this.promiseRunning) {
      while (this.storedPromises.length > 0) {
        if (this.order === 'normal') {
          this.promiseRunning = this.storedPromises.shift();
        } else if (this.order === 'reverse') {
          this.promiseRunning = this.storedPromises.pop();
        }

        try {
          var _this$promiseRunning;

          if ((_this$promiseRunning = this.promiseRunning) !== null && _this$promiseRunning !== void 0 && _this$promiseRunning.promiseFunction) {
            await this.promiseRunning.promiseFunction();
          }
        } finally {
          this.promiseRunning = undefined;
        }
      }
    }
  }
  /**
   * @inheritdoc
   */


  addPromise(key, promiseFunction) {
    const promiseObject = {
      key
    };
    const promise = new Promise((resolve, reject) => {
      promiseObject.promiseFunction = () => {
        return promiseFunction().then(resolve, reject);
      };
    });
    promiseObject.promise = promise;
    this.storedPromises.push(promiseObject);
    this.promiseExecutor();
    return promise;
  }

}

exports.SigaaPromiseStack = SigaaPromiseStack;
