// sigaa-api/dist/helpers/sigaa-request-stack.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaRequestStack = void 0;

require("source-map-support/register");

var _sigaaPromiseStack = require("./sigaa-promise-stack");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @category Internal
 */
class SigaaRequestStack {
  constructor() {
    _defineProperty(this, "_stacks", {});
  }

  createStacks(domain) {
    const newRequestStacks = {
      noCookie: new _sigaaPromiseStack.SigaaPromiseStack('reverse'),
      get: new _sigaaPromiseStack.SigaaPromiseStack('reverse'),
      post: new _sigaaPromiseStack.SigaaPromiseStack('reverse')
    };
    this._stacks[domain] = newRequestStacks;
    return this._stacks[domain];
  }
  /**
   * @inheritdoc
   */


  getStacksByDomain(domain) {
    return this._stacks[domain] || this.createStacks(domain);
  }
  /**
   * @inheritdoc
   */


  close() {
    for (const domain of Object.keys(this._stacks)) {
      this.getStacksByDomain(domain).noCookie.flush();
      this.getStacksByDomain(domain).get.flush();
      this.getStacksByDomain(domain).post.flush();
    }

    this._stacks = {};
  }

}

exports.SigaaRequestStack = SigaaRequestStack;
